contracts/UniswapSlidingWindowOracle.sol

```diff
@@ -0,0 +1,111 @@
+    uint public immutable periodSize;
+
+    // mapping from pair address to a list of price observations of that pair
+    mapping(address => Observation[]) public pairObservations;
+
+    constructor(uint windowSize_, uint8 granularity_) public {
+        require(granularity_ > 1, 'SlidingWindowOracle: GRANULARITY');
+        require(
+            (periodSize = windowSize_ / granularity_) * granularity_ == windowSize_,
+            'SlidingWindowOracle: WINDOW_NOT_EVENLY_DIVISIBLE'
+        );
+        windowSize = windowSize_;
+        granularity = granularity_;
+    }
+
+    // returns the index of the observation corresponding to the given timestamp
+    function observationIndexOf(uint timestamp) public view returns (uint8 index) {
+        uint epochPeriod = timestamp / periodSize;
+        return uint8(epochPeriod % granularity);
+    }
+
+    // returns the observation from the oldest epoch (at the beginning of the window) relative to the current time
+    function getFirstObservationInWindow(address pair) private view returns (Observation storage firstObservation) {
+        uint8 observationIndex = observationIndexOf(block.timestamp);
+        // no overflow issue. if observationIndex + 1 overflows, result is still zero.
+        uint8 firstObservationIndex = (observationIndex + 1) % granularity;
+        firstObservation = pairObservations[pair][firstObservationIndex];
+    }
+
+    // update the cumulative price for the observation at the current timestamp. each observation is updated at most
+    // once per epoch period.
+    function update(address pair) external {
+
+        // populate the array with empty observations (first call only)
+        for (uint i = pairObservations[pair].length; i < granularity; i++) {
+            pairObservations[pair].push();
+        }
+
+        // get the observation for the current period
+        uint8 observationIndex = observationIndexOf(block.timestamp);
+        Observation storage observation = pairObservations[pair][observationIndex];
+
+        // we only want to commit updates once per period (i.e. windowSize / granularity)
+        uint timeElapsed = block.timestamp - observation.timestamp;
+        if (timeElapsed > periodSize) {
+            (uint price0Cumulative, uint price1Cumulative,) = UniswapV2OracleLibrary.currentCumulativePrices(pair);
+            observation.timestamp = block.timestamp;
+            observation.price0Cumulative = price0Cumulative;
+            observation.price1Cumulative = price1Cumulative;
+        }
+    }
+
+    // returns the amount out corresponding to the amount in for a given token using the moving average over the time
+    // range [now - [windowSize, windowSize - periodSize * 2], now]
+    // update must have been called for the bucket corresponding to timestamp `now - windowSize`
+    function getPrice(address pair) external view returns (uint price0Average, uint price1Average) {
```

Comment:
hm, so `price0Average` is `reserve1/reserve0`, and `price1Average` is `reserve0/reserve1` [link](https://github.com/Uniswap/uniswap-v2-core/blob/4dd59067c76dea4a0e8e4bfdda41877a6b16dedc/contracts/UniswapV2Pair.sol#L79). we'll have to use price1 or price2 average, depending on which reserve is WETH and which is the token (Uniswap gives it a deterministic but arbitrary [order](https://github.com/Uniswap/uniswap-v2-core/blob/4dd59067c76dea4a0e8e4bfdda41877a6b16dedc/contracts/UniswapV2Factory.sol#L25)). this is something we should probably build into this layer, not have other layers worry about this

Reply:
actually, itd save us storage to only store the correct price in the `Observation` struct too

Reply:
Good point. For most of the pairs it's price0, only eth-zrx is price1.